HNSW tuning guide (#3684)
opensearch-project/project-website: _posts/2025-02-28-a-practical-guide-for-selecting-HNSW-hyperparameters.md

| [NYTimes](https://archive.ics.uci.edu/dataset/164/bag+of+wordsD199572657/)        | 256          | 290,000      | 10,000      |         100 | Angular       | BoW                                            | Language, news article        |                                                      
| [NYTimes](https://archive.ics.uci.edu/dataset/164/bag+of+wordsD199572657/)        | 16           | 290,000      | 10,000      |         100 | Angular       | BoW                                            | Language, news article        |                                                    
| [SIFT](http://corpus-texmex.irisa.fr/)           | 128          | 1,000,000    | 10,000      |         100 | Euclidean     | SIFT descriptors                               | Image                         |                                          
| [SIFT](https://github.com/erikbern/ann-benchmarks/tree/main)            | 256          | 1,000,000    | 10,000      |         100 | Hamming       | SIFT descriptors                               | Image                         |                                          

> | [Last.fm](http://millionsongdataset.com/lastfm/)        | 65           | 292,385      | 50,000      |         100 | Inner product | Matric Factorization                           | Song recommendation           |                                                   

Reviewer: @huibinshen Do you mean "matrix factorization"?

An alternative is **transfer learning**, where knowledge from optimizing one dataset is applied to another. This approach helps identify configurations that approximate optimal results while maintaining efficiency [3, 4].  

### Recommended HNSW configurations  

In the next section, we introduce a method for selecting HNSW configurations. Based on this approach, we provide **five precomputed configurations** that progressively increase graph density. These configurations cover a range of trade-offs between search quality and speed across different datasets.  

To optimize your search performance, you can **evaluate these five configurations sequentially**, stopping once recall meets your requirements. Because the configurations are ordered by increasing search quality, testing them in this order is likely to yield better search quality with each step:  

```python
{'M': 16,  'efConstruction': 128, 'efSearch': 32}
{'M': 32,  'efConstruction': 128, 'efSearch': 32}
{'M': 16,  'efConstruction': 128, 'efSearch': 128}
{'M': 64,  'efConstruction': 128, 'efSearch': 128}
{'M': 128, 'efConstruction': 256, 'efSearch': 256}
```

## Portfolio learning for HNSW  

> Portfolio learning [2, 3, 4] selects a set of complementary configurations so that at least one performs well on average when evaluated across different scenarios. Applying this approach to HNSW, we aim to identify a set of configurations that balance recall and query throughput.  

Reviewer 2: ```suggestion
Portfolio learning [2, 3, 4] selects a set of complementary configurations so that at least one performs well on average when evaluated across different scenarios. Applying this approach to HNSW, we aimed to identify a set of configurations that balance recall and query throughput.  
```

> To achieve this, we use 15 vector search datasets spanning various modalities, embedding models, and distance functions. For each dataset, we establish ground truth by computing the top 10 nearest neighbors for every query in the test set using exact k-NN search.  

Reviewer 2: ```suggestion
To achieve this, we used 15 vector search datasets spanning various modalities, embedding models, and distance functions. For each dataset, we established ground truth by computing the top 10 nearest neighbors for every query in the test set using exact k-NN search.  
```

> 

Reviewer 2: The table should be referenced in the text preceding it.

Reviewer: Changed intro paragraph.

| [GIST](http://corpus-texmex.irisa.fr/)         | 960          | 1,000,000    | 1,000       |         100 | Euclidean     | GIST descriptors, INRIA C implementation       | Image                         |                                                        
| [MS MARCO](https://microsoft.github.io/msmarco/)        | 384          | 1,000,000    | 50,000      |         100 | Euclidean     | MiniLLM                                        | Language, Question answering  |                                                           
| [openai-dbpedia](https://huggingface.co/datasets/Qdrant/dbpedia-entities-openai3-text-embedding-3-large-1536-1M) | 1,536        | 950,000      | 50,000      |         100 | Euclidean     | text-embedding-3-large                         | Language, DBPedia             | 

For each dataset, we evaluated a grid of 80 HNSW configurations based on the following search space:

```python
search_space = {
    "M": [8, 16, 32, 64, 128],
    "efConstruction": [32, 64, 128, 256],
    "efSearch": [32, 64, 128, 256]
}
```

> For these experiments, we used an OpenSearch 2.15 cluster with three cluster manager nodes and six data nodes, each running on an `r6g.4xlarge.search` instance. We evaluated test vectors in batches of 100 and recorded query throughput and recall@10 for each HNSW configuration. In the next section, we'll introduce the algorithm used to learn the portfolio.  

Reviewer 2: ```suggestion
For these experiments, we used an OpenSearch 2.15 cluster with three cluster manager nodes and six data nodes, each running on an `r6g.4xlarge.search` instance. We evaluated test vectors in batches of 100 and recorded query throughput and recall@10 for each HNSW configuration. In the next section, we introduce the algorithm used to learn the portfolio.  
```

### Method  

To capture different trade-offs between recall and throughput, we used a simple linearization approach, assigning values between 0 and 1, inclusive, to both recall and throughput. Given a specific weighting, we identified the configuration that maximizes the linearized object using the following four steps:  

1. **Normalize recall and throughput** – Apply min-max scaling within each dataset so that recall and throughput values are comparable.  

> 2. **Compute weighted metric** – Combine the normalized recall and throughput using the assigned weights into a new weighted metric.  

Reviewer 2: ```suggestion
2. **Compute weighted metric** – Using the assigned weights, combine the normalized recall and throughput into a new weighted metric.  
```

3. **Average across datasets** – Calculate the average weighted metric across datasets.  
4. **Select the best configuration** – Identify the configuration that maximizes the average weighted metric.  

> The following figure illustrates the algorithm using an example with two datasets and three configurations.

Reviewer 2: "The following figure illustrates the use of the algorithm with an example containing two datasets and three configurations"?

Reviewer: I think the original is clearer.

![Learning methods](/assets/media/blog-images/2025-02-28-a-pratical-guide-for-selecting-HNSW-hyperparameters/hnsw-portfolio-learn.png){:class="img-centered"}

We used the following weighting profiles for recall and throughput. We did not assign higher weights to throughput because most applications prioritize achieving good recall before optimizing for throughput.

|              |   0 |   1 |   2 |   3 |   4 |
|:-------------|----:|----:|----:|----:|----:|
| `w_recall`     | 0.9 | 0.8 | 0.7 | 0.6 | 0.5 |
| `w_throughput` | 0.1 | 0.2 | 0.3 | 0.4 | 0.5 |

## Evaluation  

> We evaluated our method in two scenarios:  

Reviewer 2: ```suggestion
We evaluated our method using two scenarios:  
```

> 1) **Leave-one-out evaluation** – One of the 15 datasets is used as the test dataset while the remaining datasets serve as the training set.  

Reviewer 2: ```suggestion
1. **Leave-one-out evaluation** – One of the 15 datasets is used as the test dataset while the remaining datasets serve as the training set.  
```

> 2) **Deployment evaluation** – All 15 datasets are used for training, and the method is tested on four additional datasets using a new embedding model, [Cohere-embed-english-v3](https://huggingface.co/Cohere/Cohere-embed-english-v3.0), which was not part of the training set.  

Reviewer 2: ```suggestion
2. **Deployment evaluation** – All 15 datasets are used for training, and the method is tested on 4 additional datasets using a new embedding model, [Cohere-embed-english-v3](https://huggingface.co/Cohere/Cohere-embed-english-v3.0), that was not part of the training set.  
```

> The first scenario mimics cross-validation in machine learning, while the second simulates an evaluation in which the complete training dataset is used for model deployment.  

Reviewer 2: ```suggestion
The first scenario mimics cross-validation in ML, while the second simulates an evaluation in which the complete training dataset is used for model deployment.  
```

### Leave-one-out evaluation  

> For this evaluation, we determined the ground-truth configurations under different weightings by applying our method to the test dataset. We then compared these with the predicted configurations derived from the training datasets using the same method.  

Reviewer 2: ```suggestion
For this evaluation, we determined the ground truth configurations under different weightings by applying our method to the test dataset. We then compared these with the predicted configurations derived from the training datasets using the same method.  
```

> We calculated the mean absolute error (MAE) between the predicted and ground-truth configurations for normalized (min-max scaled) recall and throughput. The following bar plot shows the average MAE across all 15 datasets in the leave-one-out evaluation:  

Reviewer 2: ```suggestion
We calculated the mean absolute error (MAE) between the predicted and ground truth configurations for normalized (min-max scaled) recall and throughput. The following bar plot shows the average MAE across all 15 datasets in the leave-one-out evaluation.  
```

![MAE compared with ground truth](/assets/media/blog-images/2025-02-28-a-pratical-guide-for-selecting-HNSW-hyperparameters/mae.png){:class="img-centered"}  

The results show that the average MAEs for normalized recall are below 0.1. For context, if recall values range from 0.5 to 0.95 on a dataset, an MAE of 0.1 translates to a raw recall difference of only 0.045. This suggests that the predicted configurations closely match the ground-truth configurations, particularly for high-recall weightings.  

The MAEs for throughput are larger, likely because throughput measurements tend to be noisier than recall measurements. However, the MAEs decrease when higher weight is assigned to throughput.  

### Deployment evaluation  

> For this evaluation, we applied our method to the 15 training datasets and tested the resulting configurations on three datasets using the Cohere-embed-english-v3 embedding model. Our goal was to ensure that the learned configurations align with the Pareto front, representing different trade-offs between recall and throughput.  

Reviewer 2: ```suggestion
For this evaluation, we applied our method to the 15 training datasets and tested the resulting configurations on 3 datasets using the Cohere-embed-english-v3 embedding model. Our goal was to ensure that the learned configurations align with the Pareto front, representing different trade-offs between recall and throughput.  
```

> The following plot depicts the recall and throughput for the learned configurations in different colors, with other configurations displayed in gray.  

Reviewer 2: Nit: On line 127 we use "bar plot" instead of just "plot". Should we also use "bar plot" here?

Reviewer: We can't because this one is not a bar plot.

> The following plot depicts the recall and throughput for the learned configurations in different colors, with other configurations displayed in gray.  

Reviewer: ```suggestion
The following bar plot depicts the recall and throughput for the learned configurations in different colors, with other configurations displayed in gray.  
```

![Trade-off of the 5 configurations](/assets/media/blog-images/2025-02-28-a-pratical-guide-for-selecting-HNSW-hyperparameters/tradeoff.png){:class="img-centered"}  

> The results show that the five selected configurations effectively cover the high-recall and high-throughput regions. Since we did not assign high weights to throughput, the learned configurations do not extend into the low-recall, high-throughput area.

Reviewer 2: ```suggestion
The results show that the five selected configurations effectively cover the high-recall and high-throughput regions. Because we did not assign high weights to throughput, the learned configurations do not extend into the low-recall, high-throughput area.
```

## How to apply the configurations in OpenSearch

> To try out these configurations, first create an index. You must specify the index build parameters when creating the index because the parameters are not dynamic:

Reviewer 2: ```suggestion
To try these configurations, first create an index. You must specify the index build parameters when creating the index because the parameters are not dynamic:
```

        method="GET", url=f"/_plugins/_knn/warmup/{index_name}?pretty"
    )
    stats = client.indices.stats(index=index_name, metric="store")
    index_size_in_bytes = stats["indices"][index_name]["total"]["store"][
        "size_in_bytes"
    ]
    graph_mem_in_kb = get_graph_memory(client)

    time.sleep(random_delay(lower_time_limit=5, upper_time_limit=10))

    logger.info(f"{index_name}: Query indexes")
    pred_inds, search_time = query_index(config, index_name, client, query_vectors, k)

    groundtruth_topk_neighbors = [v[:k] for v in neighbors]
    recall = compute_recall(np.array(groundtruth_topk_neighbors), np.array(pred_inds))
    logger.info(f"{index_name}: Recall {recall}")

    config.update(
        {
            f"recall@{k}": recall,
            "search_time": search_time,
            "search_throughput": len(query_vectors) / search_time,
            "index_size_in_bytes": index_size_in_bytes,
            "graph_mem_in_kb": graph_mem_in_kb,
        }
    )

    # clean up
    delete_one_index(client, index_name)
    logger.info(f"Clean up done, finishing evaluation.")
    return config
```

#### Define the variables for your experiments

> For your experiments, define the following variables:

Reviewer 2: ```suggestion
Define the following variables for your experiments:
```

- OpenSearch domain and engine

> - AWS region and AWS profile

Reviewer 2: ```suggestion
- AWS Region and AWS profile
```
